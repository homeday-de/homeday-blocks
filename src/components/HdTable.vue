<template>
  <table class="hd-table" :class="{ [align]: true, noWrap, fixed }">
    <thead>
      <tr>
        <th v-for="(th, i) in header" :key="`header-${i}`">{{ th }}</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(tr, i) in body" :key="`row-${i}`">
        <td v-for="(value, i) in tr" :key="`cell-${i}`">
          <component v-if="isComponent(value)" :is="value.component" v-bind="value.props"></component>
          <span v-else>{{ value }}</span>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
export default {
  name: 'HdTable',
  props: {
    header: {
      type: Array,
      required: true,
      validator(value) {
        return value.length > 0;
      },
    },
    body: {
      type: Array,
      required: true,
      validator(value) {
        return value.length > 0;
      },
    },
    align: {
      type: String,
      default: 'center',
    },
    noWrap: Boolean,
    fixed: Boolean,
  },
  methods: {
    isComponent(value) {
      return typeof value === 'object' && value.component !== undefined;
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'hd-blocks/styles/mixins.scss';

.hd-table {
  margin-top: $stack-s;
  @include font('text-xsmall');
  border-collapse: collapse;
  width: 100%;
  z-index: -1;
  text-align: center;
  th, td {
    padding: $inset-s;
    border-top: 1px solid $heather;
    color: $nevada;
    font-weight: 600;
    vertical-align: middle;
    position: relative;
  }
  th {
    border-top: none;
    color: $cello;
    font-size: 13px;
    font-weight: 700;
    vertical-align: middle;
  }
  tr:nth-child(even) {
    background-color: $wild-sand;
    td:after {
      background-color: $wild-sand;
    }
  }
  &.fixed {
    table-layout: fixed;
  }
  &.center {
    text-align: center;
  }
  &.noWrap {
    td {
      white-space: nowrap;
      overflow: hidden;
    }
    td:after {
      content: '';
      width: $inset-s;
      height: 100%;
      position: absolute;
      right: 0;
      top: 0;
      background: linear-gradient(to left, white, rgba(255, 255, 255, 0));
    }
    tr:nth-child(even) td:after {
      background: linear-gradient(to left, rgba(245, 245, 245, 1), rgba(245, 245, 245, 0));
    }
  }
  &.left {
    th, td {
      text-align: left;
    }
  }
}
</style>
