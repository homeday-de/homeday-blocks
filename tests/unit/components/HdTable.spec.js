import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdTable from '@/components/HdTable.vue';

const headerData = ['a', 'b', 'c'];
const bodyData = [
  [1, 2, 3],
  [4, 5, 6],
];

const wrapperBuilder = wrapperFactoryBuilder(HdTable, {
  props: {
    header: headerData,
    body: bodyData,
  },
});

describe('HdTable', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('creates a <table>', () => {
    const wrapper = wrapperBuilder();
    const table = wrapper.find('table');

    expect(table.exists()).toBe(true);
  });

  describe('methods', () => {
    describe('isComponent', () => {
      it('false for random object', () => {
        const wrapper = wrapperBuilder();
        const object = {};

        expect(wrapper.vm.isComponent(object)).toBe(false);
      });

      it('true for object with property component', () => {
        const wrapper = wrapperBuilder();
        const object = { component: HdTable };

        expect(wrapper.vm.isComponent(object)).toBe(true);
      });
    });
  });
});
