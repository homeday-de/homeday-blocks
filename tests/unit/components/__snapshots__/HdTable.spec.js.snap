// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdTable renders component 1`] = `
<table class="hd-table center">
  <thead>
    <tr>
      <th>a</th>
      <th>b</th>
      <th>c</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><span>1</span></td>
      <td><span>2</span></td>
      <td><span>3</span></td>
    </tr>
    <tr>
      <td><span>4</span></td>
      <td><span>5</span></td>
      <td><span>6</span></td>
    </tr>
  </tbody>
</table>
`;
